pkg/api/api.go

```diff
@@ -43,97 +43,138 @@ type podInfo struct {
+		a.log.Debug("API readiness disabled")
+	}
 }
 
-// Start runs the API.
-func (a *API) Start() {
-	a.log.Debug("Starting API")
-
-	go a.Run()
+// SetConfig sets the current dynamic configuration.
+func (a *API) SetConfig(cfg *dynamic.Configuration) {
+	a.configuration.Set(cfg)
 }
 
-// Run wraps the listenAndServe method.
-func (a *API) Run() {
-	a.log.Error(http.ListenAndServe(fmt.Sprintf("%s:%d", a.apiHost, a.apiPort), a.router))
+// SetTopology sets the current topology.
+func (a *API) SetTopology(topo *topology.Topology) {
+	a.topology.Set(topo)
 }
 
-// EnableReadiness enables the readiness flag in the API.
-func (a *API) EnableReadiness() {
-	if !a.readiness {
-		a.log.Debug("API readiness enabled")
+// getCurrentConfiguration returns the current configuration.
+func (a *API) getCurrentConfiguration(w http.ResponseWriter, _ *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+
+	if err := json.NewEncoder(w).Encode(a.configuration.Get()); err != nil {
+		a.log.Errorf("Unable to serialize dynamic configuration: %v", err)
+		a.writeErrorResponse(w, nil, http.StatusInternalServerError)
 
-		a.readiness = true
+		return
 	}
 }
 
-// getCurrentConfiguration returns the current configuration.
-func (a *API) getCurrentConfiguration(w http.ResponseWriter, r *http.Request) {
+// getCurrentTopology returns the current topology.
+func (a *API) getCurrentTopology(w http.ResponseWriter, _ *http.Request) {
 	w.Header().Set("Content-Type", "application/json")
 
-	if err := json.NewEncoder(w).Encode(a.currentConfiguration.Get()); err != nil {
-		a.log.Error(err)
+	if err := json.NewEncoder(w).Encode(a.topology.Get()); err != nil {
+		a.log.Errorf("Unable to serialize topology: %v", err)
+		a.writeErrorResponse(w, nil, http.StatusInternalServerError)
+
+		return
 	}
 }
 
 // getReadiness returns the current readiness value, and sets the status code to 500 if not ready.
-func (a *API) getReadiness(w http.ResponseWriter, r *http.Request) {
-	if !a.readiness {
-		w.WriteHeader(http.StatusInternalServerError)
+func (a *API) getReadiness(w http.ResponseWriter, _ *http.Request) {
+	isReady, _ := a.readiness.Get().(bool)
+	if !isReady {
+		a.writeErrorResponse(w, nil, http.StatusInternalServerError)
+		return
 	}
 
 	w.Header().Set("Content-Type", "application/json")
 
-	if err := json.NewEncoder(w).Encode(a.readiness); err != nil {
-		a.log.Error(err)
+	if err := json.NewEncoder(w).Encode(isReady); err != nil {
+		a.log.Errorf("Unable to serialize readiness: %v", err)
+		a.writeErrorResponse(w, nil, http.StatusInternalServerError)
+
+		return
 	}
 }
 
 // getMeshNodes returns a list of mesh nodes visible from the controller, and some basic readiness info.
-func (a *API) getMeshNodes(w http.ResponseWriter, r *http.Request) {
-	podInfoList := []podInfo{}
-
-	sel := labels.Everything()
+func (a *API) getMeshNodes(w http.ResponseWriter, _ *http.Request) {
+	// Make sure it returns an empty array and not "null".
+	podInfoList := make([]podInfo, 0)
```

Comment:
Should be initialized after the pods retrieval.

```suggestion
	podInfoList := make([]podInfo, len(podList))
```